src/compiler/checker.ts

```diff
@@ -10343,6 +10343,10 @@ namespace ts {
             if (!type) {
                 return undefined;
             }
+            // If the contextual signature has fewer parameters than the function expression, do not use it
+            if (isFunctionExpressionOrArrowFunction(node) && isAritySmaller(type, node)) {
+                return undefined;
+            }

```

Comment:
This should be the last check you do. That way you only have to deal with a single signature as everything else will have been disqualified.